.travis.yml

```diff
@@ -26,32 +26,52 @@ install:
 script:
   - bash .travis.sh
 
-env:
-  global:
-    - MAKEFLAGS="-j 2"
-    - CCACHE_SLOPPINESS=pch_defines,time_macros
+# We first do the compile stage specified below, then the matrix expansion specified after.
+stages:
+  - compile
+  - test
 
-# gcc is too slow and we have a time limit in Travis CI: selective builds.
-matrix:
+# Compile stage without building examples/tests to populate the caches.
+jobs:
   include:
-  - compiler: gcc
-    os: linux
-    env: GTSAM_ALLOW_DEPRECATED_SINCE_V4=OFF GTSAM_BUILD_TESTS=OFF GTSAM_BUILD_UNSTABLE=ON GTSAM_BUILD_EXAMPLES_ALWAYS=ON
-  - compiler: gcc
-    os: linux
-    env: GTSAM_ALLOW_DEPRECATED_SINCE_V4=OFF GTSAM_BUILD_TESTS=ON GTSAM_BUILD_UNSTABLE=OFF GTSAM_BUILD_EXAMPLES_ALWAYS=OFF  # gcc too slow for all tests
-  - compiler: gcc
+# on Mac, GCC
+  - stage: compile
+    os: osx
+    compiler: gcc
+    env: GTSAM_ALLOW_DEPRECATED_SINCE_V4=OFF GTSAM_BUILD_TESTS=OFF GTSAM_BUILD_UNSTABLE=ON GTSAM_BUILD_EXAMPLES_ALWAYS=OFF
+# on Mac, CLANG
+  - stage: compile
+    os: osx
+    compiler: clang
+    env: GTSAM_ALLOW_DEPRECATED_SINCE_V4=OFF GTSAM_BUILD_TESTS=OFF GTSAM_BUILD_UNSTABLE=ON GTSAM_BUILD_EXAMPLES_ALWAYS=OFF
+# on Linux, GCC
+  - stage: compile
     os: linux
-    env: GTSAM_ALLOW_DEPRECATED_SINCE_V4=OFF GTSAM_BUILD_TESTS=OFF GTSAM_BUILD_UNSTABLE=ON GTSAM_BUILD_EXAMPLES_ALWAYS=ON GCC_VERSION="8"
-  - compiler: clang
+    compiler: gcc
+    env: GTSAM_ALLOW_DEPRECATED_SINCE_V4=OFF GTSAM_BUILD_TESTS=OFF GTSAM_BUILD_UNSTABLE=ON GTSAM_BUILD_EXAMPLES_ALWAYS=OFF
+# on Linux, CLANG
+  - stage: compile
     os: linux
-    env: GTSAM_ALLOW_DEPRECATED_SINCE_V4=OFF GTSAM_BUILD_TESTS=ON GTSAM_BUILD_UNSTABLE=ON GTSAM_BUILD_EXAMPLES_ALWAYS=ON
-  - compiler: clang
+    compiler: clang
+    env: GTSAM_ALLOW_DEPRECATED_SINCE_V4=OFF GTSAM_BUILD_TESTS=OFF GTSAM_BUILD_UNSTABLE=ON GTSAM_BUILD_EXAMPLES_ALWAYS=OFF
+# on Linux, with deprecated ON to make sure that path still compiles
+  - stage: compile
     os: linux
-    env: GTSAM_ALLOW_DEPRECATED_SINCE_V4=ON GTSAM_BUILD_TESTS=ON GTSAM_BUILD_UNSTABLE=ON GTSAM_BUILD_EXAMPLES_ALWAYS=ON
-# - compiler: gcc
-#   os: osx
-#   env: GTSAM_ALLOW_DEPRECATED_SINCE_V4=OFF GTSAM_BUILD_TESTS=OFF GTSAM_BUILD_UNSTABLE=ON GTSAM_BUILD_EXAMPLES_ALWAYS=ON
-  - compiler: clang
-    os: osx
-    env: GTSAM_ALLOW_DEPRECATED_SINCE_V4=OFF GTSAM_BUILD_TESTS=ON GTSAM_BUILD_UNSTABLE=ON GTSAM_BUILD_EXAMPLES_ALWAYS=OFF 
+    compiler: clang
+    env: GTSAM_ALLOW_DEPRECATED_SINCE_V4=ON GTSAM_BUILD_TESTS=OFF GTSAM_BUILD_UNSTABLE=ON GTSAM_BUILD_EXAMPLES_ALWAYS=OFF
+
+# Matrix configuration:
+os:
+  - osx
+  - linux
+compiler:
+  - gcc
+  - clang
+env:
+  global:
+    - MAKEFLAGS="-j 2"
+    - CCACHE_SLOPPINESS=pch_defines,time_macros
+  matrix:
+    - GTSAM_ALLOW_DEPRECATED_SINCE_V4=OFF GTSAM_BUILD_TESTS=OFF GTSAM_BUILD_UNSTABLE=ON GTSAM_BUILD_EXAMPLES_ALWAYS=ON
+    - GTSAM_ALLOW_DEPRECATED_SINCE_V4=OFF GTSAM_BUILD_TESTS=ON GTSAM_BUILD_UNSTABLE=ON GTSAM_BUILD_EXAMPLES_ALWAYS=OFF
```

Comment:
I *think* that this "matrix" block could (and should?) be removed.
At least, it could be removed to check whether the list of generated jobs changes; if not, it would be clear it has no effects.

See: https://github.com/travis-ci/docs-travis-ci-com/issues/1500

Reply:
No, I think it works as expected: it generates the eight builds: 2 platforms, 2 compilers, 2 environment settings